cabal-install/Distribution/Client/CmdRepl.hs

```diff
@@ -100,62 +212,223 @@ replCommand = Client.installCommand {
-
-    buildCtx' <-
-      runProjectPreBuildPhase verbosity baseCtx $ \elaboratedPlan -> do
-
-            when (buildSettingOnlyDeps (buildSettings baseCtx)) $
-              die' verbosity $ "The repl command does not support '--only-dependencies'. "
-                 ++ "You may wish to use 'build --only-dependencies' and then "
-                 ++ "use 'repl'."
-
-            -- Interpret the targets on the command line as repl targets
-            -- (as opposed to say build or haddock targets).
-            targets <- either (reportTargetProblems verbosity) return
-                     $ resolveTargets
-                         selectPackageTargets
-                         selectComponentTarget
-                         TargetProblemCommon
-                         elaboratedPlan
-                         Nothing
-                         targetSelectors
-
-            -- Reject multiple targets, or at least targets in different
-            -- components. It is ok to have two module/file targets in the
-            -- same component, but not two that live in different components.
-            when (Set.size (distinctTargetComponents targets) > 1) $
-              reportTargetProblems verbosity
-                [TargetProblemMultipleTargets targets]
-
-            let elaboratedPlan' = pruneInstallPlanToTargets
-                                    TargetActionRepl
-                                    targets
-                                    elaboratedPlan
-            return (elaboratedPlan', targets)
+    let
+      onlySpecified = fromFlagOrDefault False (envOnlySpecified envFlags)
+      with    = withProject    cliConfig verbosity targetStrings
+      without = withoutProject cliConfig verbosity targetStrings
+    
+    (baseCtx, targetSelectors, finalizer) <- 
+      if onlySpecified
+        then 
+          without
+        else
+          catch with
+            $ \case
+              (BadPackageLocations prov locs) 
+                | prov == Set.singleton Implicit
+                , let 
+                  isGlobErr (BadLocGlobEmptyMatch _) = True
+                  isGlobErr _ = False
+                , any isGlobErr locs ->
+                  without
+              err -> throwIO err
+
+    when (buildSettingOnlyDeps (buildSettings baseCtx)) $
+      die' verbosity $ "The repl command does not support '--only-dependencies'. "
+          ++ "You may wish to use 'build --only-dependencies' and then "
+          ++ "use 'repl'."
+
+    baseCtx' <- if null (envPackages envFlags)
+      then return baseCtx
+      else
+        -- Unfortunately, the best way to do this is to let the normal solver
+        -- help us resolve the targets, but that isn't ideal for performance,
+        -- especially in the no-project case.
+        withInstallPlan (lessVerbose verbosity) baseCtx $ \elaboratedPlan -> do
+          -- Interpret the targets on the command line as repl targets
```

Comment:
The `targets <- ...` and the `Reject multiple targets...` bits look like they could be moved to a helper function to avoid copy-paste.